source/elements/oneMKL/source/domains/spblas/operations/spmv.rst

```diff
@@ -0,0 +1,378 @@
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           std::size_t                                &temp_buffer_size,
+       );
+
+       void spmv_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           sycl::buffer<std::uint8_t, 1>              workspace,
+       );
+
+       sycl::event spmv_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           void*                                      workspace,
+           const std::vector<sycl::event>             &dependencies = {},
+       );
+
+       sycl::event spmv(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           const std::vector<sycl::event>             &dependencies = {},
+       );
+
+   }
+
+.. container:: section
+
+   .. rubric:: Notes
+
+   - ``spmv_buffer_size`` and ``spmv_optimize`` must be called at least once before ``spmv``
+     with the same arguments. ``spmv`` can then be called multiple times.
+   - The data of the handles can be reset-ed before each call to ``spmv``.
```

Comment:
Probably leave this particular line out for now and keep it vague. We'll be able to make a more informed decision when we have an implementation later on.

Reply:
I don't understand why this should be more vague? Don't we want to guarantee that each operation can be called multiples times with different data?

Reply:
The goal of the optimize functions is to allow to run the compute function multiple times at a cheaper cost. Surely the user must be allowed to change something between each call to the compute function. It would be good to specify what can be changed, or do you think it should be backend-specific?

From cuSPARSE:
> The function cusparseSpMM_preprocess() can be called before cusparseSpMM to speedup the actual computation. It is useful when cusparseSpMM is called multiple times with the same sparsity pattern (matA). The values of the matrices (matA, matB, matC) can change arbitrarily. It provides performance advantages is used with CUSPARSE_SPMM_CSR_ALG1 or CUSPARSE_SPMM_CSR_ALG3. For all other formats and algorithms have no effect.

For spmv, the documentation is not as explicit but I think works the same way:
> Using cusparseSpMV_preprocess() helps improve performance of cusparseSpMV() in CSR. It is beneficial when we need to run cusparseSpMV() multiple times with a same matrix (cusparseSpMV_preprocess() is executed only once).

I think for spsv you are right, updating the sparse matrix requires the function `cusparseSpSV_updateMatrix`. From a little experiment I ran, the user can update the dense vector X between calls to `cusparseSpSV_solve`.

I think in conclusion we could say that the values from sparse matrices can be updated as long as the sparsity pattern stays the same for `spmm` and `spmv` only. I expect that for every operator the user is allowed to update the data of dense handles. What do you think?

Reply:
I have updated this https://github.com/uxlfoundation/oneAPI-spec/pull/522/commits/51e406894e237358e9609a226c0cff1f275e0afe. Let me know if that makes sense to you.

Reply:
Changing dense matrices/vectors constants, and or `transpose` types between the operations keeping the matrix fixed is definitely allowed. We definitely can't have the sparsity pattern changing between repeated calls, because that is entirely useless and almost equivalent to creating a new handle, so the sparsity pattern must remain fixed; now or in the future, we cannot have that change.

The last remaining thing is swapping out values array with new values while keeping the pattern same. Ideally, we want to be able to get to a stage where we allow this. Oftentimes, however, it is very difficult to allow that because of optimizations we store in the handle, operation descriptor, or the workspace. If users change the values, then we must discard or reperform optimizations we may have performed. This means the API updating the values must also pass in the operation descriptors and/or workspaces. If we can't have that, then any optimizations we may be storing in the workspace need to be values-agnostic, which can make the optimizations "less optimum". At the very least, I'm afraid the Intel oneMKL backend won't be able to support swapping out values in the near future.

Even though we'd like users to have this flexibility, it's best to disallow this for now, and if possible in the future, find a way to allow this.

Reply:
Great, thanks!